src/mlpack/core/data/split_data.hpp

```diff
@@ -61,20 +64,28 @@ void Split(const arma::Mat<T>& input,
   trainLabel.set_size(trainSize);
   testLabel.set_size(testSize);
 
-  const arma::Col<size_t> order =
-      arma::shuffle(arma::linspace<arma::Col<size_t>>(0, input.n_cols - 1,
-                                                      input.n_cols));
-
-  for (size_t i = 0; i != trainSize; ++i)
+  if (shuffleData)
   {
-    trainData.col(i) = input.col(order[i]);
-    trainLabel(i) = inputLabel(order[i]);
+    arma::Col<size_t> order;
+    order = arma::shuffle(arma::linspace<arma::Col<size_t>>(0,
+        input.n_cols - 1, input.n_cols));
+    for (size_t i = 0; i != trainSize; ++i)
+    {
+      trainData.col(i) = input.col(order[i]);
+      trainLabel(i) = inputLabel(order[i]);
+    }
+    for (size_t i = 0; i != testSize; ++i)
+    {
+      testData.col(i) = input.col(order[i + trainSize]);
+      testLabel(i) = inputLabel(order[i + trainSize]);
+    }
```

Comment:
Now that we require Armadillo 8.400.0, which has non-contiguous subviews, I think we can do what I've written below, which is really cool!

```suggestion
    trainData = input.cols(order.subvec(1, trainSize));
    trainLabel = inputLabel.subvec(order.subvec(1, trainSize));

    testData = input.cols(order.subvec(trainSize, input.n_cols - 1));
    testLabel = inputLabels.subvec(order.subvec(trainSize, input.n_cols - 1));
```

Be sure to double-check that code... I'm not 100% sure it works...

Also, I didn't suggest the change for the other one or two places where code like this occurs, so if this works, could you apply the idea there too? :)

Reply:
There was slight issues, changed it accordingly